internal/olm/client/client.go

```diff
@@ -214,7 +216,53 @@ func (c Client) DoCSVWait(ctx context.Context, key types.NamespacedName) error {
 		return curPhase == olmapiv1alpha1.CSVPhaseSucceeded, nil
 	}
 
-	return wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	err := wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		depCheckErr := c.getDeploymentStatus(ctx, key, csv)
+		if depCheckErr != nil {
+			return fmt.Errorf("failed to run operator: %v %v", err, depCheckErr)
+		}
+	}
+
+	return err
+}
+
+// Get Deployment status
+
+func (c Client) getDeploymentStatus(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+	dep := &appsv1.Deployment{}
+	var depName string
+	for _, depSpec := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		if depSpec.Spec.Selector.MatchLabels["control-plane"] == "controller-manager" {
+			depName = depSpec.Name
+		}
+	}
+	depKey := types.NamespacedName{
+		Namespace: key.Namespace,
+		Name:      depName,
+	}
+	if depErr := c.KubeClient.Get(ctx, depKey, dep); depErr != nil {
+		return fmt.Errorf("failed to run operator: Error getting deployment: %v", depErr)
+	}
+	for _, s := range dep.Status.Conditions {
+		if s.Type == "Available" && s.Status == "True" {
+			return fmt.Errorf("failed to run operator: Deployment is available, but CSV pending")
+		}
```

Comment:
Why does Available and True mean that the CSV is pending? When the CSV is not pending what would the Condition be? Available & False?

Reply:
What would be helpful to me is maybe a comment either above the loop or as part of the godoc for `getDeploymentStatus` explaining it.

Reply:
I agree , the logic is not straight forward . We check the deployment status , only when CSV is failing, as we want to dig for the actual reason. So, Deployment being "Available=true", means that, there is no deployment issue as such. But we are here as the CSV is failing.

Reply:
I don't think we need to explicitly check for the conditions being in their expected status.

Can we just check for the unexpected conditions (e.g. Available=False), and if we don't find anything unexpected, we just fall through and return nil from this function?

Then in the future if we encounter the generic timeout error message, that tells us there's probably another check we can add for some other reason that the CSV never succeeded.